core/iwasm/common/wasm_c_api.c

```diff
@@ -320,6 +320,7 @@ wasm_engine_new()
         singleton_engine =
             wasm_engine_new_internal(Alloc_With_System_Allocator, NULL);
     }
+    singleton_engine->ref_count++;
```

Comment:
Hi, currently wasm-c-api APIs are not thread-safe, if to make them thread-safe, there might be more work to do. If you insist on it, could you please add a global lock, initialize it and lock/unlock it in the operations related to singleton:
after L314, define: `static korp_mutex c_api_global_lock;`
and here: use `os_mutex_init(&c_api_global_lock)` to initialize it, and check return value

We will merge #1010 to ensure the vector related operations be safe-thread, but if there is other issue occurring, we need to fix it.

Reply:
Actually introducing locks could turn WAMR potentially useless for my project because one of the threads accessing the C API is real-time. That RT thread only issues calls to `wasm_func_call()` that's why I write *potentially*.

- Is the lock only needed for initialization/teardown of stuff?  in other words, is it guaranteed that `wasm_func_call()` never needs to lock?  I see `wasm_trap_new()` also touches the engine.

- My project (for example) does not need concurrency, because only a single thread accesses the engine at a given time. One of the threads inits/deinits the engine and another uses it (calls Wasm functions), both things don't happen simultaneously. Does it make sense to make MT support optional in WAMR?

Reply:
> Actually introducing locks could turn WAMR potentially useless for my project because one of the threads accessing the C API is real-time. That RT thread only issues calls to `wasm_func_call()` that's why I write _potentially_.
> 
> * Is the lock only needed for initialization/teardown of stuff?  in other words, is it guaranteed that `wasm_func_call()` never needs to lock?  I see `wasm_trap_new()` also touches the engine.

Yes, wasm_func_call() doesn't need lock, for multi-thread, the wasm module inst should be duplicated, and each thread accesses its own module inst. The `wasm_trap_new` only reads engine->stores->num_elems, seems no need to add lock for it.

> * My project (for example) does not need concurrency, because only a single thread accesses the engine at a given time. One of the threads inits/deinits the engine and another uses it (calls Wasm functions), both things don't happen simultaneously. Does it make sense to make MT support optional in WAMR?

Although your project doesn't require multi-threading support, your patch really makes it look like multi-threading is supported, other developers may think so and apply it in their projects. I have a discussion with lum1n0us, we prefer to add locks for single engine and vectors so as to enhance the security. And that should have no impact on `wasm_func_call`. We will merge your PR firstly, and then modify PR #1010 to add lock for single engine, is it OK for you? Thanks.

Reply:
That's sounds perfect. Yes I agree this patch alone can make things confusing as it is only needed to handle the rare AABB init/deinit case and does not have anything to do with MT.

Thank you for having a look at this.